README.md
# libsass for Dart

![libsass+dart](http://i.imgur.com/kh8QIRG.png)

A Dart VM Native Extension for libsass.

This package provides Dart language bindings for the [libsass](http://libsass.org/) library, the C version of the popular stylesheet preprocessor, Sass.

## Build Instructions

Because this library contains a Dart VM Native Extension, you will need to compile the extension before using the library. Currently, the build script only supports Linux and Mac (untested):

```shell
git clone https://github.com/oddrationale/dart_libsass.git
cd dart_libsass
dart tool/build.dart
```

This script will pull the libsass submodule, compile libsass, and compile the extenstion. Support for Windows is coming soon!

## Command Line

This package also provides a `sassc` compatible executable:

```shell
dart bin/sassd.dart input.scss output.css
```

CHANGELOG.md
# Changelog

## 0.0.1

- Initial version
- Based on libsass v3.1.0
- Wrapped libsass functions in Dart classes
- A `sassc` compatible executable: `sassd`



tool/build.sh
#!/bin/bash

set -e

cd $(dirname $0)
PROJECT_ROOT=$(pwd)/..
DART_BIN=$(which dart)
DART_SDK_DIR=$(dirname $DART_BIN)/..
PLATFORM="$(uname -s)"
DART_VERSION=$(dart --version 2>&1)
LIBSASS_SRC=$PROJECT_ROOT/lib/src/libsass
case "$DART_VERSION" in
  (*32*)
    MACOS_ARCH="i386"
    LINUX_ARCH="32"
    ;;
  (*64*)
    MACOS_ARCH="x86_64"
    LINUX_ARCH="64"
    ;;
  (*)
    echo Unsupported dart architecture $DART_VERSION.  Exiting ... >&2
    exit 3
    ;;
esac

# see https://www.dartlang.org/articles/native-extensions-for-standalone-dart-vm/
cd $PROJECT_ROOT
pub install

# Check if libsass submodule exists, if not pull submodules
if [ ! -e "$LIBSASS_SRC/Readme.md" ]; then
  echo Pulling libsass submodule
  git submodule update --init --recursive
fi

# Compile libsass
if [ ! -e "$LIBSASS_SRC/lib/libsass.a" ]; then
  echo Compiling libsass
  cd $LIBSASS_SRC
  make -j5 
fi

cd $PROJECT_ROOT/lib/src
echo Building libsass for platform $PLATFORM/$MACOS_ARCH
case "$PLATFORM" in
#  (Darwin)
#    g++ -fPIC -I $DART_SDK_DIR/include -c sync_socket_extension.cc -arch $MACOS_ARCH
#    gcc -shared -Wl,-install_name,libsync_socket_extension.dylib,-undefined,dynamic_lookup,-arch,$MACOS_ARCH -o \
#      ../libsync_socket_extension.dylib sync_socket_extension.o
#    ;;
  (Linux)
    g++ -fPIC -I$DART_SDK_DIR -Ilibsass/lib -c sass_extension.cc -lsass -m$LINUX_ARCH
    gcc -shared -fPIC -Wl,-soname,libsass_extension.so -o \
      ../libsass_extension.so sass_extension.o libsass/*.o -lc
    ;;
  (*)
    echo Unsupported platform $PLATFORM.  Exiting ... >&2
    exit 3
    ;;
esac


lib/libsass.dart
/// Dart class-based wrappers for the libsass C/C++ Sass compiler.
///
/// This library implements the libsass C/C++ Sass compiler using Dart classes.
/// The classes closely follow the functions laid out in the `sass_context.h`
/// file in the original libsass library. For further documentation, please
/// refer to the
/// [libsass implementation guide](https://github.com/sass/libsass/wiki).
library libsass;

import 'dart-ext:sass_extension';

abstract class Sass {
  /// The version of the compiled libsass C/C++ library.
  static String libsassVersion() native "libsass_version";
}

class SassOptions {
  int _handle;

  SassOptions() {
    _handle = _sassMakeOptions();
    _peerRegister(this, _handle);
  }

  SassOptions.fromHandle(this._handle);

  int get handle => _handle;

  int get precision => _sassOptionGetPrecision(_handle);
  int get outputStyle => _sassOptionGetOutputStyle(_handle);
  bool get sourceComments => _sassOptionGetSourceComments(_handle);
  bool get sourceMapEmbed => _sassOptionGetSourceMapEmbed(_handle);
  bool get sourceMapContents => _sassOptionGetSourceMapContents(_handle);
  bool get omitSourceMapUrl => _sassOptionGetOmitSourceMapUrl(_handle);
  bool get isIndentedSyntaxSrc => _sassOptionGetIsIndentedSyntaxSrc(_handle);
  String get inputPath => _sassOptionGetInputPath(_handle);
  String get outputPath => _sassOptionGetOutputPath(_handle);
  String get includePath => _sassOptionGetIncludePath(_handle);
  String get sourceMapFile => _sassOptionGetSourceMapFile(_handle);

  void set precision(int precision) =>
      _sassOptionSetPrecision(_handle, precision);
  void set outputStyle(int outputStyle) =>
      _sassOptionSetOutputStyle(_handle, outputStyle);
  void set sourceComments(bool sourceComments) =>
      _sassOptionSetSourceComments(_handle, sourceComments);
  void set sourceMapEmbed(bool sourceMapEmbed) =>
      _sassOptionSetSourceMapEmbed(_handle, sourceMapEmbed);
  void set sourceMapContents(bool sourceMapContents) =>
      _sassOptionSetSourceMapContents(_handle, sourceMapContents);
  void set omitSourceMapUrl(bool omitSourceMapUrl) =>
      _sassOptionSetOmitSourceMapUrl(_handle, omitSourceMapUrl);
  void set isIndentedSyntaxSrc(bool isIndentedSyntaxSrc) =>
      _sassOptionSetIsIndentedSyntaxSrc(_handle, isIndentedSyntaxSrc);
  void set inputPath(String inputPath) =>
      _sassOptionSetInputPath(_handle, inputPath);
  void set outputPath(String outputPath) =>
      _sassOptionSetOutputPath(_handle, outputPath);
  void set includePath(String includePath) =>
      _sassOptionSetIncludePath(_handle, includePath);
  void set sourceMapFile(String sourceMapFile) =>
      _sassOptionSetSourceMapFile(_handle, sourceMapFile);

  // Native functions
  int _sassMakeOptions() native "sass_make_options";
  void _peerRegister(Object object, int handle) native "peer_register";

  int _sassOptionGetPrecision(int handle) native "sass_option_get_precision";
  int _sassOptionGetOutputStyle(int handle) native
      "sass_option_get_output_style";
  bool _sassOptionGetSourceComments(int handle) native
      "sass_option_get_source_comments";
  bool _sassOptionGetSourceMapEmbed(int handle) native
      "sass_option_get_source_map_embed";
  bool _sassOptionGetSourceMapContents(int handle) native
      "sass_option_get_source_map_contents";
  bool _sassOptionGetOmitSourceMapUrl(int handle) native
      "sass_option_get_omit_source_map_url";
  bool _sassOptionGetIsIndentedSyntaxSrc(int handle) native
      "sass_option_get_is_indented_syntax_src";
  String _sassOptionGetInputPath(int handle) native
      "sass_option_get_input_path";
  String _sassOptionGetOutputPath(int handle) native
      "sass_option_get_output_path";
  String _sassOptionGetIncludePath(int handle) native
      "sass_option_get_include_path";
  String _sassOptionGetSourceMapFile(int handle) native
      "sass_option_get_source_map_file";

  void _sassOptionSetPrecision(int handle, int precision) native
      "sass_option_set_precision";
  void _sassOptionSetOutputStyle(int handle, int outputStyle) native
      "sass_option_set_output_style";
  void _sassOptionSetSourceComments(int handle, bool sourceComments) native
      "sass_option_set_source_comments";
  void _sassOptionSetSourceMapEmbed(int handle, bool sourceMapEmbed) native
      "sass_option_set_source_map_embed";
  void _sassOptionSetSourceMapContents(int handle,
      bool sourceMapContents) native "sass_option_set_source_map_contents";
  void _sassOptionSetOmitSourceMapUrl(int handle, bool omitSourceMapUrl) native
      "sass_option_set_omit_source_map_url";
  void _sassOptionSetIsIndentedSyntaxSrc(int handle,
      bool isIndentedSyntaxSrc) native "sass_option_set_is_indented_syntax_src";
  void _sassOptionSetInputPath(int handle, String inputPath) native
      "sass_option_set_input_path";
  void _sassOptionSetOutputPath(int handle, String outputPath) native
      "sass_option_set_output_path";
  void _sassOptionSetIncludePath(int handle, String includePath) native
      "sass_option_set_include_path";
  void _sassOptionSetSourceMapFile(int handle, String sourceMapFile) native
      "sass_option_set_source_map_file";
}

abstract class SassContext {
  int _handle;
  SassOptions _options;

  int compile();

  SassOptions get options => _options;
  void set options(SassOptions options);

  String get outputString => _sassContextGetOutputString(_handle);
  int get errorStatus => _sassContextGetErrorStatus(_handle);
  String get errorJson => _sassContextGetErrorJson(_handle);
  String get sourceMapString => _sassContextGetSourceMapString(_handle);
  String get includedFiles => _sassContextGetIncludedFiles(_handle);

  // Native functions
  void _peerRegister(Object object, int handle) native "peer_register";

  String _sassContextGetOutputString(int handle) native
      "sass_context_get_output_string";
  int _sassContextGetErrorStatus(int handle) native
      "sass_context_get_error_status";
  String _sassContextGetErrorJson(int handle) native
      "sass_context_get_error_json";
  String _sassContextGetSourceMapString(int handle) native
      "sass_context_get_source_map_string";
  String _sassContextGetIncludedFiles(int handle) native
      "sass_context_get_included_files";
}

class SassFileContext extends SassContext {

  SassFileContext(String inputPath) {
    _handle = _sassMakeFileContext(inputPath);
    _peerRegister(this, _handle);
    _options = new SassOptions.fromHandle(_sassFileContextGetOptions(_handle));
  }

  int compile() => _sassCompileFileContext(_handle);

  void set options(SassOptions options) {
    _sassFileContextSetOptions(_handle, options.handle);
    _options = new SassOptions.fromHandle(_sassFileContextGetOptions(_handle));
  }

  // Native functions
  int _sassMakeFileContext(String inputPath) native "sass_make_file_context";
  int _sassCompileFileContext(int handle) native "sass_compile_file_context";
  int _sassFileContextGetOptions(int handle) native
      "sass_file_context_get_options";
  void _sassFileContextSetOptions(int handle, int options) native
      "sass_file_context_set_options";
}

class SassDataContext extends SassContext {

  SassDataContext(String sourceString) {
    _handle = _sassMakeDataContext(sourceString);
    _peerRegister(this, _handle);
    _options = new SassOptions.fromHandle(_sassDataContextGetOptions(_handle));
  }

  int compile() => _sassCompileDataContext(_handle);

  void set options(SassOptions options) {
    _sassDataContextSetOptions(_handle, options.handle);
    _options = new SassOptions.fromHandle(_sassDataContextGetOptions(_handle));
  }

  // Native functions
  int _sassMakeDataContext(String sourceString) native "sass_make_data_context";
  int _sassCompileDataContext(int handle) native "sass_compile_data_context";
  int _sassDataContextGetOptions(int handle) native
      "sass_data_context_get_options";
  void _sassDataContextSetOptions(int handle, int options) native
      "sass_data_context_set_options";
}


tool/build.dart
import 'dart:async';
import 'dart:io';

import 'package:cli_util/cli_util.dart' as cli_util;
import 'package:path/path.dart' as path;

String makeJobsArgument = '-j${(Platform.numberOfProcessors+1).toString()}';
String dartSdkArch;

Directory dartSdkDir = cli_util.getSdkDir().absolute;
Directory projectRoot = new Directory.fromUri(Platform.script).parent.parent;
Directory extensionSourceDir = new Directory(path.join(projectRoot.path, 'lib', 'src'));
Directory libsassSourceDir = new Directory(path.join(extensionSourceDir.path, 'libsass'));
Directory libsassLibDir = new Directory(path.join(libsassSourceDir.path, 'lib'));

File dartBin = new File(path.join(dartSdkDir.absolute.path, 'bin', 'dart'));
File pubBin = new File(path.join(dartSdkDir.absolute.path, 'bin', 'pub'));

ProcessArgs pubGetProcess = new ProcessArgs(pubBin.path, ['install'], wD: projectRoot);
ProcessArgs gitSubmoduleUpdate = new ProcessArgs('git', ['submodule', 'update', '--init', '--recursive'], wD: projectRoot);
ProcessArgs makeLibSassProcess = new ProcessArgs('make', [makeJobsArgument], wD: libsassSourceDir);

void main() {
  // stderr, really? Does not seem right, but that's where it's printed
  String dartVersionString = Process.runSync(dartBin.path, ['--version']).stderr;
  print('''
ENV: 
  dartSdkVersion     = $dartVersionString
  dartSdkDir         = $dartSdkDir
  projectRoot        = $projectRoot
  extensionSourceDir = $extensionSourceDir
  libsassSourceDir   = $libsassSourceDir
  libsassLibDir      = $libsassLibDir
''');

  projectRootPrepare()
    ..then((_) {
    if(Platform.isLinux) {
      if(new RegExp(r"linux_x64").firstMatch(dartVersionString) != null) {
        dartSdkArch = '64';
      } else {
        dartSdkArch = '32';
      }
      extensionCompileNix();
    } else if(Platform.isMacOS) {
      if(new RegExp(r"macos_x64").firstMatch(dartVersionString) != null) {
        dartSdkArch = 'x86_64';
      } else {
        dartSdkArch = 'i386';
      }
      extensionCompileNix();
    } else {
      print('unsupported dart-sdk: "$dartVersionString" located: "${dartSdkDir.path}"');
      exit(1);
    }
  })
  ..catchError((Exception _) => print('Could not prepare environment. ${_.toString()}'));
}

Future projectRootPrepare() {
  final Completer ac = new Completer();
  Directory.current = projectRoot;
  pubGetProcess.run()
    ..then((_) {
    gitSubmoduleUpdate.run().then((_) => ac.complete());
  })
  ..catchError((_) => ac.completeError(new Exception('pubGet failed!')));
  return ac.future;
}

Future extensionCompileNix() {
  Completer ac = new Completer();
  ProcessArgs makeExtensionProcessS1;
  ProcessArgs makeExtensionProcessS2;

  // Assemble Compilation Process 1
  makeExtensionProcessS1 = new ProcessArgs('g++',
  ['-fPIC',
  '-I${dartSdkDir.path}',
  '-I${libsassLibDir.path}',
  '-c', 'sass_extension.cc'],
  wD: extensionSourceDir);
  makeExtensionProcessS2 = new ProcessArgs('gcc', ['-fPIC'], wD: extensionSourceDir);

  // Find .o files in libsass source directory
  RegExp oFileReg = new RegExp(r".o$");
  List<FileSystemEntity> libsassOutFiles = libsassSourceDir.listSync(recursive: false).where((_) => oFileReg.firstMatch(_.path) != null);
  if(Platform.isLinux) {
    makeExtensionProcessS1.arguments.addAll([
        '-m$dartSdkArch',
        '-lsass'
    ]);
    makeExtensionProcessS2.arguments.addAll([
        '-shared',
        '-m$dartSdkArch',
        '-Wl,-soname,libsass_extension.so',
        '-o', '../libsass_extension.so',
        '-lc'
    ]);
  } else if(Platform.isMacOS) {
    makeExtensionProcessS1.arguments.addAll([
        '-arch', dartSdkArch
    ]);
    makeExtensionProcessS2.arguments.addAll([
        '-dynamiclib',
        '-undefined', 'dynamic_lookup',
        '-arch', dartSdkArch,
        '-Wl,-install_name,libsass_extension.dylib',
        '-o', '../libsass_extension.dylib',
        '-lc'
    ]);
  }
  // append .o files to process-arguments
  makeExtensionProcessS2.arguments.addAll(libsassOutFiles.map((_) => _.absolute.path));

  makeLibSassProcess.run()
    ..then((_) {
    makeExtensionProcessS1.run()
      ..then((_) {
      makeExtensionProcessS2.run()
        ..then((_) {
        print('Done. Everything went fine.');
        ac.complete();
      })
        ..catchError((_) => ac.completeError(new Exception('Compilation failed in stage 2. See console output.'))
      );
    })
      ..catchError((_) => ac.completeError(new Exception('Compilation failed in stage 1. See console output.')));
  })
    ..catchError((_) => ac.completeError(new Exception('Make of libsass failed. See console output.')));
  return ac.future;
}

class ProcessArgs {
  String executable;
  List<String> arguments;
  Directory wD;

  ProcessArgs(this.executable, this.arguments, {this.wD: null});

  Future run() {
    final Completer ac = new Completer();
    if (wD != null) Directory.current = wD;
    print('# Running: "$executable ${arguments.join(' ')}" in "${Directory.current.path}"');
    Process.start(executable, arguments, runInShell: true).then((Process _) {
      _.stderr.transform(SYSTEM_ENCODING.decoder).listen(print);
      _.stdout.transform(SYSTEM_ENCODING.decoder).listen(print);
      _.exitCode.then((_) {
        if (_ == 0) {
          ac.complete(_);
          print('# Success!');
        } else ac.completeError(new Exception('Exit code was: $_'));
      });
    });
    return ac.future;
  }
}